### PR TITLE
boards/bluepill-stm32f030c8: add STM32F030C8 based bluepill board

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

STM32F030C8 based Bluepill

This board is sold by several vendors on [AliExpress](https://aliexpress.com/item/4001132135566.html), [Amazon](https://www.amazon.de/ILS-STM32F030C8T6-Coreboard-Systemplatine-Development/dp/B07ZP1C7L7/) and [ebay](https://www.ebay.de/itm/STM32F030C8T6-Core-Board-System-Board-STM32-F0-ARM-Development-Board/123949949952).
It does not have name, but since it comes in the same form-factor and color as the Bluepill line of boards, I think it makes sense to use that name.

Flashing is a bit cumbersome as I always had to press the reset button in time when OpenOCD tries to connect to the CPU.
There is no dedicated reset pin exposed on the board.

### CODE REVIEW

File: boards/bluepill-stm32f030c8/doc.txt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+![STM32F030C8 based Bluepill](https://camo.githubusercontent.com/f33ec5f4068e1dcc4c549dbe3dc70d2ed784c19abf1c177f9640835180f88005/68747470733a2f2f696d616765732d6e612e73736c2d696d616765732d616d617a6f6e2e636f6d2f696d616765732f492f3631474765592532424c306a4c2e5f41435f534c313030305f2e6a7067)
+
+### MCU
+| MCU        | STM32F030C8T6         |
+|:---------- |:--------------------- |
+| Family     | ARM Cortex-M0         |
+| Vendor     | ST Microelectronics   |
+| RAM        | 8 kiB                 |
+| Flash      | 64 kiB                |
```
Comment (maribu) [inline boards/bluepill-stm32f030c8/doc.txt:20]:
```suggestion
| RAM        | 8 KiB                 |
| Flash      | 64 KiB                |
```

File: boards/bluepill-stm32f030c8/doc.txt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+| Vcc        | 2.0V - 3.6V           |
+| Datasheet  | [Datasheet](https://www.st.com/en/microcontrollers-microprocessors/stm32f030f4.html) |
+| Reference Manual | [Reference Manual](https://www.st.com/resource/en/datasheet/stm32f030f4.pdf) |
+| Programming Manual | [Programming Manual](http://www.st.com/resource/en/programming_manual/dm00051352.pdf) |
+
+## Flashing the device
+
+The STM32F030C8 based Bluepill board does not include a programmer.
+You have to connect a separate ST-Link programmer to the (SW)DIO, (SW)CLK, GND and
+NRST pins on the board.
```
Comment (maribu) [inline boards/bluepill-stm32f030c8/doc.txt:38]: This contradicts the PR description. Does it have an NRST pin, or not?

File: boards/bluepill-stm32f030c8/doc.txt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+```
+
+## Known Issues
+
+### Flashing fails
+
+The board does not expose a pin for the reset signal but wires it only to the
+reset button.
+You can work around this by pressing the reset button when OpenOCD wants to connect
+to the Blue Pill.
```
Comment (maribu) [inline boards/bluepill-stm32f030c8/doc.txt:62]: Have you tried keeping reset pressed until OpenOCD tries to connect, then release? OpenOCD tries to connect for like 2 seconds or so - so this worked really reliable for me.

**Review by maribu — COMMENTED** [review]
Code looks good. I have some comments for the doc

File: boards/bluepill-stm32f030c8/doc.txt (hunk 2 above)
Comment (benpicco) [reply boards/bluepill-stm32f030c8/doc.txt]: unfortunately not :(

**Comment by fjmolinas** [comment]
@benpicco is it possible to provide some test output?

**Comment by fjmolinas** [comment]
> @benpicco is it possible to provide some test output?

@benpicco still the only thing missing (well and a rebase now)

**Comment by maribu** [comment]
I bought some of these, so I could also test this :-)

> Flashing is a bit cumbersome as I always had to press the reset button in time when OpenOCD tries to connect to the CPU. There is no dedicated reset pin exposed on the board.

If you drop the "connect under reset" option in the OpenOCD config, it should be able to connect fine - mostly. It will not connect during WFI or when crashed, though. Maybe a reliable but cumbersome procedure is better than an non-cumbersome but unreliable.

**Comment by maribu** [comment]
Btw: Mind to also provide a default MTD config (but still letting the user to explicitly add the module)? To me, it looks like the unpopulated footprint U3 is just waiting for SPI flash to be soldered on. But this is just a suggestion.

File: boards/bluepill-stm32f030c8/include/periph_conf.h [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+#define CONFIG_BOARD_HAS_LSE        1
+#endif
+
+/* This board provides an HSE */
+#ifndef CONFIG_BOARD_HAS_HSE
+#define CONFIG_BOARD_HAS_HSE        1
+#endif
+
+#include "periph_cpu.h"
+#include "f0/cfg_clock_default.h"
```
Comment (maribu) [inline boards/bluepill-stm32f030c8/include/periph_conf.h:34]:
```suggestion
#include "clk_conf.h"
```

This needs to be updated to the clock config cleanup.

**Review by maribu — COMMENTED** [review]
see inline

**Review by maribu — COMMENTED** [review]
With this, I get I2C working.

**Comment by maribu** [comment]
Test results:

- [x] UART (only one present, works for shell interaction)
- [x] LED mapping (SAUL shell command allows controlling it)
- [x] PWM (output tested with logic analyzer and `tests/periph_pwm`)
- [x] SPI (tested `examples/default` with `USEMODULE=bme280_spi` and could read the sensor via SPI)
- [x] I2C (tested with above patches and `example/default` with `USEMODULE="bme280_i2c i2c_scan` and communication with the sensor worked reliable, also `i2c_scan` worked and my logic analyzer can decode the I2C stream without warnings)
- [X] ADC (tested with `tests/periph_adc`)

File: boards/bluepill-stm32f030c8/doc.txt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+## Known Issues
+
+### Flashing fails
+
+The board does not expose a pin for the reset signal but wires it only to the
+reset button.
+You can work around this by pressing the reset button when OpenOCD wants to connect
+to the Blue Pill, or keep it pressed until OpenOCD tries to connect.
+Hit the reset button again after flashing in order to boot the newly flashed image.
+
```
Comment (maribu) [inline boards/bluepill-stm32f030c8/doc.txt:64]:
````suggestion

Alternatively, flashing can be done using
```
make BOARD=bluepill-stm32f030c8 OPENOCD_RESET_USE_CONNECT_ASSERT_SRST=0 flash
```
This way, the debugger tries to connect without resetting. However, attaching without reset may fail with the board is crashed or in low power mode. In that case, you have to fall back to pressing the reset button manually.
````

Comment (benpicco) [reply boards/bluepill-stm32f030c8/doc.txt:64]: I can just add `OPENOCD_RESET_USE_CONNECT_ASSERT_SRST` to `Makefile.include` instead

Comment (maribu) [reply boards/bluepill-stm32f030c8/doc.txt:64]: But maybe adding a sentence to the doc that when flashing fails one should try with

```
make BOARD=bluepill-stm32f030c8 OPENOCD_RESET_USE_CONNECT_ASSERT_SRST=1 flash
```

with pressed reset button and releasing the button while OpenOCD tries to connect would still make sense. E.g. if an application crashes on startup, this is the only way to attach the debugger again.

**Review by maribu — APPROVED** [review]
ACK

File: boards/bluepill-stm32f030c8/doc.txt (hunk 5 above)
Comment (benpicco) [reply boards/bluepill-stm32f030c8/doc.txt:64]: Hm that doesn't seem to do anything

```
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : clock speed 1000 kHz
Info : STLINK V2J17S4 (API v2) VID:PID 0483:3748
Info : Target voltage: 3.220513
Info : stm32f0x.cpu: hardware has 4 breakpoints, 2 watchpoints
Info : starting gdb server for stm32f0x.cpu on 0
Info : Listening on port 41845 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f0x.cpu       hla_target little stm32f0x.cpu       running

Error: timed out while waiting for target halted
TARGET: stm32f0x.cpu - Not halted
```

Comment (maribu) [reply boards/bluepill-stm32f030c8/doc.txt:64]: did you do `make debug`, or `make flash`?

Comment (benpicco) [reply boards/bluepill-stm32f030c8/doc.txt:64]: I tried both `make BOARD=bluepill-stm32f030c8 OPENOCD_RESET_USE_CONNECT_ASSERT_SRST=0 flash` and `make BOARD=bluepill-stm32f030c8 OPENOCD_RESET_USE_CONNECT_ASSERT_SRST=1 flash`

Comment (maribu) [reply boards/bluepill-stm32f030c8/doc.txt:64]: This is a regression, maybe due to changed default behavior in OpenOCD. Anyway, the following patch fixes this:

```patch
diff --git a/dist/tools/openocd/openocd.sh b/dist/tools/openocd/openocd.sh
index 92dd062244..431a4b9fdb 100755
--- a/dist/tools/openocd/openocd.sh
+++ b/dist/tools/openocd/openocd.sh
@@ -142,6 +142,8 @@
 # Handle OPENOCD_RESET_USE_CONNECT_ASSERT_SRST
 if [ "${OPENOCD_RESET_USE_CONNECT_ASSERT_SRST}" = "1" ]; then
   OPENOCD_EXTRA_RESET_INIT+="-c 'reset_config connect_assert_srst'"
+else
+  OPENOCD_EXTRA_RESET_INIT+="-c 'reset_config none'"
 fi
 
 #
```

But maybe postpone this to a follow up PR. I think the openocd.cfg in `boards/common/blxxxpill` can also be ditched, only the one for the 128 KiB hack will still be needed.

Comment (maribu) [reply boards/bluepill-stm32f030c8/doc.txt:64]: Or maybe this always only worked for the Bluepill in the first place, due to this OpenOCD config?

Comment (benpicco) [reply boards/bluepill-stm32f030c8/doc.txt:64]: with this and `OPENOCD_RESET_USE_CONNECT_ASSERT_SRST=0 ` it is indeed working!

**Pull request merged**